astropy/table/operations.py

```diff
@@ -123,18 +57,21 @@ def _get_list_of_tables(tables):
     return tables
 
 
-def _get_out_class(tables):
+def _get_out_class(objs):
     """
-    From a list of table instances get the merged output table class.
-    This is just taken as the deepest subclass.  It is assumed that
-    `tables` is a list of at least one element and that they are all
-    Table (subclass) instances.  This doesn't handle complicated
+    From a list of input objects ``objs`` get merged output object class.
+    This is just taken as the deepest subclass. This doesn't handle complicated
     inheritance schemes.
     """
-    out_class = tables[0].__class__
-    for t in tables[1:]:
-        if issubclass(t.__class__, out_class):
-            out_class = t.__class__
+    out_class = objs[0].__class__
+    for obj in objs[1:]:
+        if issubclass(obj.__class__, out_class):
+            out_class = obj.__class__
+
+    if any(not issubclass(out_class, obj.__class__) for obj in objs):
```

Comment:
This now adds an explicit check that all the input objects have consistent inheritance.